### DISCUSSION TITLE
Interesting rounding quirks in temp data?

### DISCUSSION THREAD

**Discussion by JulianLelandBell** [meteostat/meteostat-python#126] [Q&A]
Hi all! Loving meteostat - it's been super helpful for a project I'm working on, and really easy to get started with for a mechanical-engineer-masquerading-as-programmer like me. Really appreciate the meteostat team for putting this together!

In my analyses, I ran across an interesting quirk in temperature data, which I wanted to share with the community/see if other folks have encountered. In my project, I am counting the number of widgets that are exposed to a given ambient temperature (and using meteostat data + lat/lon info about the widgets to estimate what that temperature is). I then create a histogram showing how many widgets were exposed to each temperature.

When I do this, I see something unusual. Here's what the histogram looks like if I bin by 1 deg C increments:

![image](https://github.com/meteostat/meteostat-python/assets/7558603/8a033c08-17ba-4752-a641-c1e63597ab75)

Here's what it looks like if I bin by 0.5 deg C increments:

![image](https://github.com/meteostat/meteostat-python/assets/7558603/30ebe34c-d01e-4568-9171-40141ba88b6a)

And here's what it looks like if I bin by 0.1 deg C increments:

![image](https://github.com/meteostat/meteostat-python/assets/7558603/2d7b202f-a31b-4e9d-b369-99c35c5054e3)

What I think is weird here: why are there dips in the histogram right before each even multiple of 5 deg C (at least in the 1C and 0.5C binned plots)?

I've gone through a range of example elements in my (very simple) widget-counting code, and verified that this is working as expected. I've also tested this at a range of different bin sizes, and it looks like it goes away above 1 deg C bins So, I *think* what I'm seeing is a) some artifact in the way that meteostat temperature data is measured (e.g., a substantial fraction of stations only report to the nearest 0.2 deg C, or all stations round up, or something similar); and/or b) something in the way that matplotlib creates stair plots (I'm not actually doing a histogram).

Anyway - I can't post code or data for a variety of reasons, but would love to hear if anyone has any ideas why I'm seeing what I'm seeing!

**Comment by denis-bz** [comment]
No idea if this is your problem, but `np.round` (and afaik stdC) rounds int + 0.5 to the nearest EVEN value --
```
np.round( 0.5 ) == 0
np.round( 1.5 ) == 2 ...
```
So if data has lots of values at exactly int + 0.5, round-to-even is bad. Try

```
import numpy as np

def roundnear( x ):
    return np.floor( x + 0.5 )

grow = 4/3
x = np.arange( 10.0 )
print( "x:", x )

for roundf in [np.round, roundnear]:
    down = roundf( x / grow )
    up = roundf( down * grow )
    diff = up - x
    print( "%g %-10s: err av %-4.2g  %s " % (
            grow, roundf.__name__, diff.mean(), diff ))
```

But how to test if `round` is a problem -- monkey-patch `around` in .../numpy/core/fromnumeric.py ?? yech
